### PR TITLE
export: Temporal spectrum and projection export to PNG, TIFF, and CSV

### PR DESCRIPTION
**Pull request by amuskens**
First draft

### CODE REVIEW

File: src/odemis/gui/comp/viewport.py [hunk 1]
```diff
@@ -1103,7 +1103,7 @@ def _on_new_data(self, data):
         Called when a new data is available (in a live stream)
         data (1D DataArray)
         """
-        if data.size:
+        if data is not None:
```
Comment (pieleric) [inline src/odemis/gui/comp/viewport.py]: I've also fixed that in my PR about the viewport. I'd suggest you drop it here.

File: src/odemis/acq/stream/_projection.py [hunk 2]
```diff
@@ -927,6 +927,8 @@ def _updateImage(self):
             data = self._computeSpec()
             if data is not None:
                 self.image.value = self._project2RGB(data)
+            else:
```
Comment (pieleric) [inline src/odemis/acq/stream/_projection.py]: Some of these changes might conflict with my PR... or not. So let's keep them as-is, and we'll see when my PR is merged.

File: src/odemis/util/spectrum.py [hunk 3]
```diff
@@ -55,7 +55,11 @@ def get_wavelength_per_pixel(da):
         wl = da.metadata[model.MD_WL_LIST]
         # check dimension of data
         dims = da.metadata.get(model.MD_DIMS, "CTZYX"[-da.ndim:])
-        if len(dims) == 3 and dims == "YXC" and da.shape[2] in (3, 4):  # RGB?
+        if len(da.shape) == 2:
+            # This is a hack to make the line spectrum select the right dimension and report
```
Comment (pieleric) [inline src/odemis/util/spectrum.py]: This is for raw line spectrum? In this case, the MD_DIMS should be "XC", and the rest should work normally.
=> Fix the .projectAsRaw() in the stream, and no need to update the function here.

File: src/odemis/gui/comp/viewport.py (hunk 1 above)
Comment (amuskens) [reply src/odemis/gui/comp/viewport.py]: Fixed what exactly?

File: src/odemis/dataio/csv.py [hunk 4]
```diff
@@ -69,32 +67,76 @@ def export(filename, data):
...
+                time_range = [t * 1e12 for t in time_range]
+                unit = "ps"
+            logging.debug("Exporting time spectrum data to CSV")
+
+            if unit == "ps":
+                # turn range to nm
+                time_tuples = [(s, d) for s, d in zip(time_range, data)]
+                headers = ['# Time (ps)', 'intensity']
+            else:
+                logging.info("Exporting time spectrum without time list information")
```
Comment (pieleric) [inline src/odemis/dataio/csv.py]: "time spectrum" -> chronogram

File: src/odemis/dataio/csv.py [hunk 5]
```diff
@@ -69,32 +67,76 @@ def export(filename, data):
...
 
+            spectrum_range, unit = spectrum.get_spectrum_range(data)
+            if unit == "m":
+                spectrum_range = [s * 1e9 for s in spectrum_range]
+                unit = "nm"
+
             # attach wavelength as first column
-            wavelength_lin = numpy.array(spectrum_range)
-            qz_masked = numpy.append(wavelength_lin.reshape(data.shape[0], 1), data, axis=1)
+            # wavelength_lin = numpy.array(spectrum_range)
```
Comment (pieleric) [inline src/odemis/dataio/csv.py]: Drop all the commented lines.

File: src/odemis/dataio/csv.py [hunk 6]
```diff
@@ -69,32 +67,76 @@ def export(filename, data):
...
-            distance_lin.shape = (1, distance_lin.shape[0])
-            distance_lin = numpy.append([[0]], distance_lin, axis=1)
-            qz_masked = numpy.append(distance_lin, qz_masked, axis=0)
-            data = model.DataArray(qz_masked, data.metadata)
+            distance_lin = numpy.linspace(0, line_length, data.shape[0])
+            distance_lin.shape = (distance_lin.shape[0], 1)
+            data = numpy.append(distance_lin, data, axis=1)
+            # data = model.DataArray(qz_masked, data.metadata)
 
             # Data should be in the form of (Y+1, X+1), with the first row and column the
```
Comment (pieleric) [inline src/odemis/dataio/csv.py]: Please update that comment. In particular, `Y+1,X+1` -> `X,C+1`.

File: src/odemis/dataio/csv.py [hunk 7]
```diff
@@ -69,32 +67,76 @@ def export(filename, data):
...
+        spectrum_range, unit_c = spectrum.get_spectrum_range(data)
+        if unit_c == "m":
+            spectrum_range = [s * 1e9 for s in spectrum_range]
+            unit_c = "nm"
+        time_range, unit_t = spectrum.get_time_range(data)
+        if unit_t == "s":
+            time_range = [t * 1e12 for t in time_range]
+            unit_t = "ps"
+
+        headers = ["time(" + unit_t + ")\wavelength(" + unit_c + ")"] + spectrum_range
```
Comment (pieleric) [inline src/odemis/dataio/csv.py]: Please write `"\\wavelength"` for safety.

File: src/odemis/dataio/csv.py [hunk 8]
```diff
@@ -112,7 +154,10 @@ def export(filename, data):
...
 
             # Set the 'header' in the 0,0 element
-            first_row = ['theta\phi[rad]'] + [d for d in data[0, 1:]]
+            first_row = ['theta\\phi[rad]'] + [d for d in data[0, 1:]]
             csv_writer.writerow(first_row)
             # dump the array
             csv_writer.writerows(data[1:, :])
+            
+    else:
+        raise IOError("Unknown acquisition type of data to be exported as CSV")
```
Comment (pieleric) [inline src/odemis/dataio/csv.py]: In such case, a `ValueError` is more appropriate (as it's a sign that the function input is incorrect).

File: src/odemis/gui/cont/export.py [hunk 9]
```diff
@@ -230,13 +231,13 @@ def export(self, export_type, raw=False):
...
+            exported_data = spectrum_to_export_data(streams[0], raw)
         elif export_type == 'spectrum-line':
-            exported_data = line_to_export_data(vp.stream, raw)
+            exported_data = line_to_export_data(streams[0], raw)
         elif export_type == 'spectrum-temporal':
-            exported_data = temporal_spectrum_to_export_data(vp.stream, raw)
+            exported_data = temporal_spectrum_to_export_data(streams[0], raw)
         elif export_type == 'spectrum-time':
-            exported_data = time_spectrum_to_export_data(vp.stream, raw)
+            exported_data = time_spectrum_to_export_data(streams[0], raw)
```
Comment (pieleric) [inline src/odemis/gui/cont/export.py]: Please rename all the "time_spectrum" to "chronogram".

**Pull request merged**